File: packages/zod/src/question.ts
import { z } from "zod";

// Section enum - 7 subtests UTBK
export const ZSection = z.enum(["PU", "PPU", "PBM", "PK", "LBI", "LBE", "PM"]);

// Solution Step schema
export const ZSolutionStep = z.object({
  order: z.number().int(),
  title: z.string(),
  content: z.string(),
});

// Base Question schema
export const ZQuestion = z.object({
  id: z.string().uuid(),
  questionBankId: z.string().uuid().nullable(),
  section: ZSection,
  subType: z.string().nullable(),

  // IRT Parameters
  difficultyIrt: z.number().nullable(),
  discrimination: z.number().nullable(),
  guessingParam: z.number().nullable(),

  // Question Content
  text: z.string(),
  optionA: z.string(),
  optionB: z.string(),
  optionC: z.string(),
  optionD: z.string(),
  optionE: z.string(),
  correctAnswer: z.enum(["A", "B", "C", "D", "E"]),

  // Explanation
  explanation: z.string().nullable(),
  explanationEn: z.string().nullable(),
  strategyTip: z.string().nullable(),
  relatedConcept: z.string().nullable(),
  solutionSteps: z.array(ZSolutionStep).nullable(),

  isActive: z.boolean(),

  // Statistics
  attemptCount: z.number().int().nullable(),
  correctRate: z.number().nullable(),
  avgTimeSeconds: z.number().int().nullable(),

  // Timestamps
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
  deletedAt: z.string().datetime().nullable(),
});

// === Request Schemas ===

// List questions query params
export const ZListQuestionsQuery = z.object({
  section: ZSection.optional(),
  page: z.coerce.number().int().min(1).default(1),
  limit: z.coerce.number().int().min(1).max(100).default(10),
});

// Get next question query params
export const ZGetNextQuestionQuery = z.object({
  section: ZSection,
});

// Get question by ID path params
export const ZGetQuestionParams = z.object({
  id: z.string().uuid(),
});

// === Response Schemas ===

// Question response (without correct answer - for practice)
export const ZQuestionResponse = z.object({
  id: z.string().uuid(),
  section: ZSection,
  subType: z.string().nullable(),
  difficultyIrt: z.number().nullable(),
  text: z.string(),
  optionA: z.string(),
  optionB: z.string(),
  optionC: z.string(),
  optionD: z.string(),
  optionE: z.string(),
  avgTimeSeconds: z.number().int().nullable(),
});

// Question detail response (with correct answer - after answering)
export const ZQuestionDetailResponse = ZQuestionResponse.extend({
  correctAnswer: z.enum(["A", "B", "C", "D", "E"]),
  explanation: z.string().nullable(),
  explanationEn: z.string().nullable(),
  strategyTip: z.string().nullable(),
  solutionSteps: z.array(ZSolutionStep).nullable(),
  relatedConcept: z.string().nullable(),
});

// Paginated questions response
export const ZQuestionListResponse = z.object({
  data: z.array(ZQuestionResponse),
  total: z.number().int(),
  page: z.number().int(),
  limit: z.number().int(),
  totalPages: z.number().int(),
});

// === Type Exports ===
export type Section = z.infer<typeof ZSection>;
export type SolutionStep = z.infer<typeof ZSolutionStep>;
export type Question = z.infer<typeof ZQuestion>;
export type ListQuestionsQuery = z.infer<typeof ZListQuestionsQuery>;
export type GetNextQuestionQuery = z.infer<typeof ZGetNextQuestionQuery>;
export type GetQuestionParams = z.infer<typeof ZGetQuestionParams>;
export type QuestionResponse = z.infer<typeof ZQuestionResponse>;
export type QuestionDetailResponse = z.infer<typeof ZQuestionDetailResponse>;
export type QuestionListResponse = z.infer<typeof ZQuestionListResponse>;

File: apps/backend/internal/model/question/question.go
package question

import (
	"time"

	"github.com/google/uuid"
	"github.com/manikandareas/genta/internal/model"
)

// Question represents a question entity
type Question struct {
	ID             uuid.UUID  `json:"id" db:"id"`
	QuestionBankID *uuid.UUID `json:"questionBankId" db:"question_bank_id"`

	Section model.Section `json:"section" db:"section"`
	SubType *string       `json:"subType" db:"sub_type"`

	// IRT Parameters
	DifficultyIRT  *float64 `json:"difficultyIrt" db:"difficulty_irt"`
	Discrimination *float64 `json:"discrimination" db:"discrimination"`
	GuessingParam  *float64 `json:"guessingParam" db:"guessing_param"`

	// Question Content
	Text          string `json:"text" db:"text"`
	OptionA       string `json:"optionA" db:"option_a"`
	OptionB       string `json:"optionB" db:"option_b"`
	OptionC       string `json:"optionC" db:"option_c"`
	OptionD       string `json:"optionD" db:"option_d"`
	OptionE       string `json:"optionE" db:"option_e"`
	CorrectAnswer string `json:"correctAnswer" db:"correct_answer"`

	// Explanation
	Explanation    *string `json:"explanation" db:"explanation"`
	ExplanationEn  *string `json:"explanationEn" db:"explanation_en"`
	StrategyTip    *string `json:"strategyTip" db:"strategy_tip"`
	RelatedConcept *string `json:"relatedConcept" db:"related_concept"`

	// Solution Steps (JSON)
	SolutionSteps *[]SolutionStep `json:"solutionSteps" db:"solution_steps"`

	IsActive bool `json:"isActive" db:"is_active"`

	// Statistics
	AttemptCount   *int     `json:"attemptCount" db:"attempt_count"`
	CorrectRate    *float64 `json:"correctRate" db:"correct_rate"`
	AvgTimeSeconds *int16   `json:"avgTimeSeconds" db:"avg_time_seconds"`

	// Timestamps
	model.BaseWithCreatedAt
	model.BaseWithUpdatedAt
	DeletedAt *time.Time `json:"deletedAt" db:"deleted_at"`
}

// SolutionStep represents a step in the solution
type SolutionStep struct {
	Order   int    `json:"order"`
	Title   string `json:"title"`
	Content string `json:"content"`
}
